### ISSUE TITLE
[knowbi/knowbi-pentaho-pdi-neo4j-output]   java.lang.OutOfMemoryError: GC overhead limit exceeded

### ISSUE THREAD

**Issue by rbortolini** [knowbi/knowbi-pentaho-pdi-neo4j-output#174]
Does version 5.0 of the Neo4JOutput plugin work with neo4j 3.5.3 Community version?

When I do a test with few records to download neo4j the transformation works with the version 5 plugin, but when I download a large volume (7 million) I have a memory error. (Image 1).

![image](https://user-images.githubusercontent.com/55597898/77191032-13afbe00-6ab9-11ea-9451-4eaf3b6a01f4.png)

The Image 2 File shows only the execution that is occurring the memory error with the error level (detailed) in the transformation, I stopped it after 30 minutes. The Process consumes almost 100% for hours and after the error.

![image](https://user-images.githubusercontent.com/55597898/77190919-d6e3c700-6ab8-11ea-90bd-345812610139.png)

Scenario with the error.
PDI: (8.2 or 9.0)
Neo4JOutput: 5.0
Neo4j: 3.5.3
Memory: "-Xms1024m" "-Xmx2048m"

When I run this same transformation with the Neo4JOutput 3.13.3 Plugin it works without problems (Image 3)
![image](https://user-images.githubusercontent.com/55597898/77190741-92f0c200-6ab8-11ea-8abe-6b5b3b6766d8.png)

Scenario without error.
PDI: (8.2 or 9.0)
Neo4JOutput: 3.13.3
Neo4j: 3.5.3
Memory: "-Xms1024m" "-Xmx2048m"

**Comment by mattcasters** [comment]
5.0 works fine against a 3.5 server.  Just make sure to turn off encryption and you're good to go.
The error is unrelated to the plugin.  You're just running out of memory.
Did you specify a very large row set size or something to that nature?  Perhaps a very large commit side on the table output step?  Something is consuming a lot of memory.

**Comment by rbortolini** [comment]
Good Morning.
The strange thing is that this same transformation with the same settings using Plugin 3.13.3 Neo4JOutput works without problems, as shown in image 3 of the previous post.

Here are the settings I am using when running the 5.0 plugin:
![image](https://user-images.githubusercontent.com/55597898/77315861-1132ab80-6ce7-11ea-8e88-f084305ae362.png)

![image](https://user-images.githubusercontent.com/55597898/77316480-4390d880-6ce8-11ea-8732-0c4279ddd933.png)


I did a new execution of the transformation, increasing the memory but it had the same behavior.

![image](https://user-images.githubusercontent.com/55597898/77316624-8357c000-6ce8-11ea-885d-9d2d0e2a59b3.png)

New scenario.
PDI: 8.2 
Neo4JOutput: 5.0
Neo4j: 3.5.3
Memory: "-Xms4096m" "-Xmx6144m"

**Comment by mattcasters** [comment]
So maybe the cypher query results are loaded into memory somehow?  It's worth checking out.

**Comment by mattcasters** [comment]
Yeah, I tracked it down to the loop over the Cypher step. It will require some deep thinking.

**Comment by mattcasters** [comment]
Please let me know if 5.0.1 pre-release fixes this for you:  https://github.com/knowbi/knowbi-pentaho-pdi-neo4j-output/releases/tag/5.0.1

**Comment by rbortolini** [comment]
With the new version Plugin 5.0.1 the transformation that was in error worked without problems. I also did the execution of the complete original process that runs some transformations in parallel and it worked without problems.
 
![image](https://user-images.githubusercontent.com/55597898/77464740-63a6c180-6de6-11ea-8ed7-9d8235a454d4.png)
 
Scenario without error.
IDP: (8.2)
Neo4JOutput: 5.0.1
Neo4j: 3.5.3
Memory: "-Xms1024m" "-Xmx2048m"


I did a new test on the version of neo4j 4.0 with Plugin 5.0.1 with the transformation that was in error and worked without problems. But when it was run complete original process that runs some transformations in parallel presented the same initial problem.

![image](https://user-images.githubusercontent.com/55597898/77465047-d6b03800-6de6-11ea-8cd3-bac1a394b04a.png)

New scenario with the error.
IDP: (8.2)
Neo4JOutput: 5.0.1
Neo4j: 4.0
Memory: "-Xms4096m" "-Xmx6144m"

**Comment by mattcasters** [comment]
Not sure how to help you with this one.  If you have something concrete to reproduce please create a new issue.

**Comment by rbortolini** [comment]
Did the problem found only affect version 3.5.3 of Neo4j or also affect other versions like 4.0?

**Comment by mattcasters** [comment]
This was unrelated to the underlying database.  It was a side effect of the Neo4j driver API change.

**Comment by rbortolini** [comment]
I was in doubt because I made two executions with the same code and the same 5.0.1 plugin, one with Neo4J db in version 3.5.3 and the other in version 4.0 and the executions had a different behavior, with memory error in version 4.0.

**Comment by rbortolini** [comment]
The Version 4.0 database was migrated from version 3.5.3 and has the same structures and volume.
